clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -0,0 +1,484 @@
+                         });
+  InitMethod = (It != CRD->methods().end()) ? *It : nullptr;
+  return InitMethod;
+}
+
+// Creates body for new OpenCL kernel. This body contains initialization of SYCL
+// kernel object fields with kernel parameters and a little bit transformed body
+// of the kernel caller function.
+static CompoundStmt *CreateOpenCLKernelBody(Sema &S,
+                                            FunctionDecl *KernelCallerFunc,
+                                            DeclContext *KernelDecl) {
+  llvm::SmallVector<Stmt *, 16> BodyStmts;
+  CXXRecordDecl *LC = getKernelObjectType(KernelCallerFunc);
+  assert(LC && "Kernel object must be available");
+  TypeSourceInfo *TSInfo = LC->isLambda() ? LC->getLambdaTypeInfo() : nullptr;
+
+  // Create a local kernel object (lambda or functor) assembled from the
+  // incoming formal parameters
+  auto KernelObjClone = VarDecl::Create(
+      S.Context, KernelDecl, SourceLocation(), SourceLocation(),
+      LC->getIdentifier(), QualType(LC->getTypeForDecl(), 0), TSInfo, SC_None);
+  Stmt *DS = new (S.Context) DeclStmt(DeclGroupRef(KernelObjClone),
+                                      SourceLocation(), SourceLocation());
+  BodyStmts.push_back(DS);
+  auto KernelObjCloneRef =
+      DeclRefExpr::Create(S.Context, NestedNameSpecifierLoc(), SourceLocation(),
+                          KernelObjClone, false, DeclarationNameInfo(),
+                          QualType(LC->getTypeForDecl(), 0), VK_LValue);
+
+  auto KernelFuncDecl = cast<FunctionDecl>(KernelDecl);
+  auto KernelFuncParam =
+      KernelFuncDecl->param_begin(); // Iterator to ParamVarDecl (VarDecl)
+  if (KernelFuncParam) {
+    llvm::SmallVector<Expr *, 16> InitExprs;
+    InitializedEntity VarEntity =
+        InitializedEntity::InitializeVariable(KernelObjClone);
+    for (auto Field : LC->fields()) {
+      // auto getExprForKernelParameter = [](Sema &S, const QualType &paramTy,
+      //                                     DeclRefExpr *DRE) {
+      //   Expr *Res = ImplicitCastExpr::Create(
+      //       S.Context, paramTy, CK_LValueToRValue, DRE, nullptr, VK_RValue);
+      //   return Res;
+      // };
+
+      // Creates Expression for special SYCL object: accessor or sampler.
+      // All special SYCL objects must have __init method, here we use it to
+      // initialize them. We create call of __init method and pass built kernel
+      // arguments as parameters to the __init method.
+      auto getExprForSpecialSYCLObj = [&](const QualType &paramTy,
+                                          FieldDecl *Field,
+                                          const CXXRecordDecl *CRD,
+                                          Expr *Base) {
+        // All special SYCL objects must have __init method
+        CXXMethodDecl *InitMethod = getInitMethod(CRD);
+        assert(InitMethod &&
+               "The accessor/sampler must have the __init method");
```

Comment:
Remove "sampler" from here?

Reply:
Sounds reasonable.
"__init method is expected"